addons/account/models/account_move.py

```diff
@@ -504,46 +2039,555 @@ def _run_post_draft_to_post(self):
+        ''' This method is used to compute 'price_total' & 'price_subtotal'.
+
+        :param price_unit:  The current price unit.
+        :param quantity:    The current quantity.
+        :param discount:    The current discount.
+        :param currency:    The line's currency.
+        :param product:     The line's product.
+        :param partner:     The line's partner.
+        :param taxes:       The applied taxes.
+        :param move_type:   The type of the move.
+        :return:            A dictionary containing 'price_subtotal' & 'price_total'.
+        '''
+        res = {}
+
+        # Compute 'price_subtotal'.
+        price_unit_wo_discount = price_unit * (1 - (discount / 100.0))
+        subtotal = quantity * price_unit_wo_discount
+
+        # Compute 'price_total'.
+        if taxes:
+            taxes_res = taxes._origin.compute_all(price_unit_wo_discount,
+                quantity=quantity, currency=currency, product=product, partner=partner, is_refund=move_type in ('out_refund', 'in_refund'))
+            res['price_subtotal'] = taxes_res['total_excluded']
+            res['price_total'] = taxes_res['total_included']
+        else:
+            res['price_total'] = res['price_subtotal'] = subtotal
+        return res
+
+    @api.model
+    def _get_computed_accounting_vals(self, price_subtotal, move_type, currency, company, date):
+        ''' This method is used to recompute the values of 'amount_currency', 'debit', 'credit' due to a change made
+        in some business fields (affecting the 'price_subtotal' field).
+
+        :param price_subtotal:  The untaxed amount.
+        :param move_type:       The type of the move.
+        :param currency:        The line's currency.
+        :param company:         The move's company.
+        :param date:            The move's date.
+        :return:                A dictionary containing 'debit', 'credit', 'amount_currency'.
+        '''
+        if move_type in ('out_refund', 'in_invoice', 'in_receipt'):
+            sign = 1
+        elif move_type in ('in_refund', 'out_invoice', 'out_receipt'):
+            sign = -1
+        else:
+            sign = 1
+        price_subtotal *= sign
+
+        if currency and currency != company.currency_id:
+            # Multi-currencies.
+            balance = currency._convert(price_subtotal, company.currency_id, company, date)
+            return {
+                'amount_currency': price_subtotal,
+                'debit': balance > 0.0 and balance or 0.0,
+                'credit': balance < 0.0 and -balance or 0.0,
+            }
+        else:
+            # Single-currency.
+            return {
+                'amount_currency': 0.0,
+                'debit': price_subtotal > 0.0 and price_subtotal or 0.0,
+                'credit': price_subtotal < 0.0 and -price_subtotal or 0.0,
+            }
+
+    @api.model
+    def _get_inversed_accounting_vals(self, quantity, discount, balance, move_type, currency, taxes):
```

Comment:
Why don't do something like this
```python
def _get_inversed_accounting_vals(self, quantity=None, discount=None, balance=None, move_type=None, currency=None, taxes=None):
    ''' doc '''
    self.ensure_one()
    quantity = quantity or self.quantity
    discount = discount or self.discount
    balance = balance or self.amount_currency
    move_type = move_type or self.move_id.type
    currency = currency or self.currency
    taxes  = taxes or self.tax_ids
```
This way you can call only ```python
line.update(line._get_inversed_accounting_vals())``` in most places (onchange and write), and give only the values that are not default. This would remove a lot of duplicated code (multiple onchanges, create and write)

The same goes for `_get_computed_business_vals` of course.

Reply:
Because I really don't like methods being both `api.model` & `api.multi`